### PR TITLE
[filetests]: attempt to compile run-command -> wat -> zkasm

### PR DESCRIPTION
**Pull request by MCJOHN974**
This PR aims to compile run command and perform correct register allocation. It uses `run-command` -> `wat` -> `zkasm` approach. WIP.

### CODE REVIEW

**Comment by MCJOHN974** [comment]
Now I finally build runnable zkasm from `.clif` file!!! Still need some refactoring to be ready to merge imo

**Comment by MCJOHN974** [comment]
Some notes before I start refactoring:
Generating preamble and postamble is out of scope of this PR
Processing error generated by `assert_eq` is out of scope of this PR

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 1]
```diff
@@ -281,68 +283,158 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
...
+        program.extend(inv);
+    }
     for foo in functions {
         program.extend(foo);
     }
     program.append(&mut postamble);
     program.join("\n")
 }
 
+fn runcommand_to_wasm(
```
Comment (mooori) [inline cranelift/filetests/src/zkasm_codegen.rs:345]: nit: `runcommand_to_wat` would be more appropriate.

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 2]
```diff
@@ -281,68 +283,158 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
...
+        func: invoke.func.clone(),
+        args: invoke.args.clone(),
+    };
+    let wat = runcommand_to_wasm(inv, cmp, expected.clone());
+    let wasm_module = wat::parse_str(wat).unwrap();
+
+    let settings = ZkasmSettings::default();
 
-    let args = invoke.args;
+    // TODO: we should not use generate_zkasm itself, but a bit changed version?
```
Comment (mooori) [inline cranelift/filetests/src/zkasm_codegen.rs:417]: Adding filetest specific fields to `ZkasmSettings` might help to achieve the desired behavior.

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 3]
```diff
@@ -281,68 +283,158 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
...
-        res.push(format!("  {} => {}", arg, registers[idx]))
-    }
-    res.push(format!("    :JMP({})", funcname));
-    // TODO: handle functions with multiple outputs
-    res.push(format!("  {} => B", expected[0]));
-    // TODO: replace with call to host function
-    res.push(format!("  CALL AWESOME ASSERT ({})", compare));
-    res
+    // TODO: will this always function_2?
+    let start_index = generated.iter().position(|r| r == "function_2:").unwrap();
```
Comment (mooori) [inline cranelift/filetests/src/zkasm_codegen.rs:426]: `generated_function` is the function that should be invoked? Relying on the generated name of that function seems brittle.

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 4]
```diff
@@ -281,68 +283,158 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
...
+    let end_index = generated.iter().rposition(|r| r == "  :JMP(RR)").unwrap();
+    let mut generated_function = generated[start_index..=end_index].to_vec();
+
+    generated_function[0] = format!("{}:", new_label);
+    // TODO: will this always function_1?
```
Comment (mooori) [inline cranelift/filetests/src/zkasm_codegen.rs:431]: What is `function_1` expected to be?

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 5]
```diff
@@ -281,68 +283,158 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
...
-    let mut res = vec![funcname];
+    let mut res = vec![format!("{}:", funcname)];
     res.append(&mut lines);
+    res.into_iter()
+        .map(|s| s.replace("label", &format!("label_{}", funcname)))
+        .collect()
+}
+
+/// Builds main for test program
+pub fn build_main(invoke_names: Vec<String>) -> Vec<String> {
```
[the remark this answers is not in the document]
Comment (MCJOHN974) [reply cranelift/filetests/src/zkasm_codegen.rs:294]: I think we can go even further and separate all function related to testing only in other file, `zkasm_testgen` or something like that.

File: cranelift/filetests/src/zkasm_codegen.rs (hunk 3 above)
Comment (MCJOHN974) [reply cranelift/filetests/src/zkasm_codegen.rs:426]: Yes, I also think it shouldn't work  by this name. However, I think in next PR with switching to `generate_preamble` with proper arguments, there will be some more reliable way to do it

File: cranelift/filetests/src/zkasm_codegen.rs (hunk 4 above)
Comment (MCJOHN974) [reply cranelift/filetests/src/zkasm_codegen.rs:431]: When compiling such `.wat` snippets, result was look like (at least for programs I was testing):
```zkasm
;; preamble
function_2:
  ;; put values to the registers
  ;; JMP(function_1)
  ;; assert_helper
  ;; JMP(RR)
```

From one side, while `.wat` code have only have 3 functions, and all of them always appear in same order, it is quite natural that the functions always named same way. From the other side, I hope to find more reliable way to do it once finishing preamble generation.

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 6]
```diff
@@ -281,68 +283,158 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
     // TODO: I believe it can be done more beautiful way
     let mut funcname = func.name.to_string();
     funcname.remove(0);
-    funcname.push(':');
-    let mut res = vec![funcname];
+    let mut res = vec![format!("{}:", funcname)];
     res.append(&mut lines);
+    res.into_iter()
+        .map(|s| s.replace("label", &format!("label_{}", funcname)))
+        .collect()
```
[the remark this answers is not in the document]
Comment (MCJOHN974) [reply cranelift/filetests/src/zkasm_codegen.rs:290]: Yes, here I replace functions names generated by `generate_zkasm` same for all compiled `run` commands. Also, it helps to make name of compiled `clif` function same in all it's calls in `run` command

File: cranelift/filetests/src/zkasm_codegen.rs (hunk 2 above)
Comment (MCJOHN974) [reply cranelift/filetests/src/zkasm_codegen.rs:424]: I hope so, but think better do it in separate PR

File: cranelift/filetests/src/zkasm_codegen.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MCJOHN974) [reply cranelift/filetests/src/zkasm_codegen.rs:301]: Created #245 about it

File: cranelift/filetests/src/zkasm_codegen.rs (hunk 4 above)
Comment (mooori) [reply cranelift/filetests/src/zkasm_codegen.rs:434]: I thinks it's quite hard to follow what happens in `compile_invocation`, even more since it relies on implementation details of codegen. The explanations from your comments would be suited for comments in the code. Ideally the comments answer these questions:

- What is the invocation compiled to? To me it's not clear from context or return type.
- How is the generated zkASM expected to look like?
- What are `function_1` and `function_2` expected to be?

File: cranelift/filetests/src/zkasm_codegen.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mooori) [reply cranelift/filetests/src/zkasm_codegen.rs:297]: I think it would be helpful to mention that in a comment. Otherwise it is hard to follow why labels are renamed (in that particular way).

**Review by mooori — COMMENTED** [review]
Could you post a command to execute zkASM filetests?

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 7]
```diff
@@ -188,6 +188,8 @@ fn fix_relocs(
             if let FinalizedRelocTarget::ExternalName(ExternalName::User(name)) = reloc.target {
                 let name = &params.user_named_funcs()[name];
                 if name.index == 0 {
+                    // Codegen line after migrating to new assert:
```
Comment (nagisa) [inline cranelift/filetests/src/zkasm_codegen.rs:191]:
```suggestion
                    // TODO(#1234): Codegen line after migrating to new assert:
```

Would be great to have a reference to something explaining what the plan for the new assert is and reference it here.

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 8]
```diff
@@ -288,68 +290,154 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
...
-    let mut res = vec![funcname];
+    let mut res = vec![format!("{}:", funcname)];
     res.append(&mut lines);
+    res.into_iter()
+        .map(|s| s.replace("label", &format!("label_{}", funcname)))
+        .collect()
+}
+
+/// Builds main for test program
+pub fn build_main(invoke_names: Vec<String>) -> Vec<String> {
```
Comment (nagisa) [inline cranelift/filetests/src/zkasm_codegen.rs:301]:
```suggestion
pub fn build_test_main(invoke_names: Vec<String>) -> Vec<String> {
```

seems like a good enough compromise for now that lets us move forward without spinning our wheels on this.

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 9]
```diff
@@ -288,68 +290,154 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
...
+        };
+        arg_types.push_str(arg_type);
+        arg_types.push_str(" ");
+
+        args_pushing.push_str(&format!("{arg_type}.const {arg}\n        "));
+    }
+    if arg_types.len() > 0 {
+        arg_types.pop();
+    }
+    // TODO: remove line with 8 whitespaces in the end of args_pushing
```
Comment (nagisa) [inline cranelift/filetests/src/zkasm_codegen.rs:377]: I would approach this in a way that makes this code read nicely and if we want to see nicely formatted wat output there're tools (wabt, wasm-tools) that will parse a messy one and output a nicely formatted version for you.

File: cranelift/filetests/src/zkasm_codegen.rs [hunk 10]
```diff
@@ -288,68 +290,154 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
...
+    }
     for foo in functions {
         program.extend(foo);
     }
     program.append(&mut postamble);
     program.join("\n")
 }
 
+fn runcommand_to_wat(invoke: Invocation, _compare: Comparison, expected: Vec<DataValue>) -> String {
+    // TODO: support different amounts of outputs
```
Comment (nagisa) [inline cranelift/filetests/src/zkasm_codegen.rs:353]: FWIW we don't need to do this for near use-cases, multi-value is an extension over the wasm-core-1 spec and nearcore runtimes don't enable it. I guess an exception to this might be if 0 outputs are expected...

File: cranelift/filetests/src/zkasm_codegen.rs (hunk 4 above)
Comment (nagisa) [reply cranelift/filetests/src/zkasm_codegen.rs:434]: This function generally seems like its trying to tip toe around our own code, which is a clear indication of an abstraction that's working against us. For instance, we appear to be parsing the output in an attempt to determine function boundaries, but nothing stops e.g. `JMP(RR)` to be used for something else than returning to a caller... LLVM for example especially likes a pattern similar to
```
    push rr
    rr <= addrof(done)
    call some_scope
some_scope:
    do_whatever_with_non-trivial control flow...
    jmp(rr)...

done:
    pop(rr)
```

within a single function in certain scenarios (e.g. to implement jumptables) and there is no reason why clift couldn't learn these tricks either.

IIRC cranelift already works on a per-function basis anyway so it might very well be worthwhile to revisit whether `generate_zkasm` makes sense if it interferes with use-cases as straightforward as this one. In fact from what I can see working around it is more code than replicating what `generate_zkasm` does in the first place... So perhaps splitting it up slightly would be a simpler approach?

File: cranelift/filetests/src/test_run_zkasm.rs [hunk 11]
```diff
@@ -56,22 +58,30 @@ impl SubTest for TestRunZkasm {
...
-        // TODO: instead of printing run program, using something like this:
-        // match zkasm_runner::run_zkasm(&zkasm_program) {
-        //     // TODO: Probably here is a good place to collect info generated by assert-hostfunction
-        //     // and somehow show it
-        //     Ok(_) => Ok(()),
-        //     Err(e) => Err(e),
-        // }
+        let execution_result = zkasm_runner::run_zkasm(&zkasm_program).unwrap();
+        match execution_result.error {
+            Some(err) => panic!("Zkasm runtime error: {}", err),
```
[the remark this answers is not in the document]
Comment (MCJOHN974) [reply cranelift/filetests/src/test_run_zkasm.rs:74]: As I understand  `error` field is used to store zkasm runtime error. And `run_zkasm` will successfully return something even if runtime error occurred during zkasm execution. Here I want to panic if this runtime error happened

**Comment by MCJOHN974** [comment]
> Could you post a command to execute zkASM filetests?

`cargo run -- test filetests/filetests/runtests/simple.clif` (or add `test run-zkasm` to header of other runtest and change path in this command). Works only from `cranelift` directory

**Comment by MCJOHN974** [comment]
Thanks all for your comments! As I understand I fixed all (or almost all) notes that block merging.

**Comment by MCJOHN974** [comment]
Also, @mooori, thanks for note that this code needs more comments. I tried add some, but also we have an issue #222 which reminds us to update documents after test infra will be ready. It will be much easier to make all this explanations easier to read in format of documentation

**Review by nagisa — APPROVED** [review]

**Comment by MCJOHN974** [comment]
Hmm, it seems that none of newest versions of `wabt` pass `cargo vet`. Should we make some changes to our `cargo-vet` policy or migrate to some other crate?

cc @nagisa

**Review by mooori — APPROVED** [review]

**Comment by nagisa** [comment]
`cargo vet` requires us to actually… vet the dependencies we pull in. This is one of those checks that I think is probably quite excessive for this project and we could consider disabling.

But that said, bytecodealliance's own https://docs.rs/wasmprinter/latest/wasmprinter/ is vetted: https://github.com/near/wasmtime/blob/c96b25f7945584eb5764d7d0d1727d68f99b363f/supply-chain/audits.toml#L280-L298

**Comment by MCJOHN974** [comment]
`wasmprinter` helps with `wasm` -> `wat` part, but don't help with `wat` -> `wasm`. Since this `wat` is just an intermediate step and can be used only for debug test infra itself, I think we can solve this `wat` formatting problem later
